File: packages/nexus-coordinator/src/nexus_coordinator/capability_store.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Capabilities gate-off-by-default store + ``@require_capability`` decorator.

Sprint 25 Phase D — D5 per CAPABILITY_TOGGLES.md design.

The store manages ``~/.sbfb/capabilities.toml``:
- TOML schema ``version = 1`` with per-capability ``[capability.X]``
  sections (``enabled``, ``enabled_at``, ``enabled_by``).
- Anti-tamper ``integrity_hash`` (SHA-256 of file excluding hash line).
- On tamper detect → fallback all-OFF + structlog warning.
- On missing file → create default all-OFF + valid hash.

The ``@require_capability`` decorator gates FastAPI endpoints — 403
when the named capability is disabled.
"""

from __future__ import annotations

import hashlib
import os
import tomllib
from dataclasses import dataclass
from datetime import datetime, timezone
from functools import wraps
from pathlib import Path
from typing import Any

import structlog
import tomli_w
from fastapi import HTTPException

_log = structlog.get_logger(__name__)

KNOWN_CAPABILITIES: frozenset[str] = frozenset(
    {
        "tool_calling",
        "streaming_bridge",
        "mcp_server_expose",
        "federation_canary",
        "rag_retrieval",
        "biometric_gate",
    }
)

CAPABILITY_DESCRIPTIONS: dict[str, str] = {
    "tool_calling": "LLM tool-calling / function-calling surface (OWASP LLM06:2025 Excessive Agency).",
    "streaming_bridge": "Streaming postMessage bridge method task_submit_streaming.",
    "mcp_server_expose": "MCP server endpoint exposed to external LLM agents.",
    "federation_canary": "Cross-jurisdiction warrant canary cosigning (FROST K-of-N).",
    "rag_retrieval": "RAG retrieval from external document sources.",
    "biometric_gate": "OS-level biometric gate for T2 loopback tier.",
}


@dataclass(frozen=True, slots=True)
class CapabilityEntry:
    enabled: bool
    enabled_at: str
    enabled_by: str


def _sbfb_dir() -> Path:
    home = os.environ.get("HOME") or os.environ.get("USERPROFILE") or ""
    if not home:
        return Path.cwd() / ".sbfb"
    return Path(home) / ".sbfb"


def default_capabilities_path() -> Path:
    return _sbfb_dir() / "capabilities.toml"


def _compute_integrity_hash(file_text: str) -> str:
    lines = file_text.splitlines(keepends=True)
    filtered = [line for line in lines if not line.startswith("integrity_hash")]
    return "sha256-" + hashlib.sha256("".join(filtered).encode()).hexdigest()


class CapabilitiesStore:
    def __init__(self, path: Path, capabilities: dict[str, CapabilityEntry]) -> None:
        self._path = path
        self._capabilities = dict(capabilities)

    @classmethod
    def load(cls, path: Path | None = None) -> CapabilitiesStore:
        if path is None:
            path = default_capabilities_path()
        if not path.exists():
            store = cls._default(path)
            store._write()
            _log.info(
                "capabilities.toml created (all OFF)",
                path=str(path),
            )
            return store

        text = path.read_text(encoding="utf-8")
        try:
            data = tomllib.loads(text)
        except Exception:
            _log.warning(
                "capabilities.toml parse error, fallback all-OFF",
                path=str(path),
            )
            return cls._default(path)

        stored_hash = data.get("integrity_hash", "")
        computed = _compute_integrity_hash(text)
        if stored_hash != computed:
            _log.warning(
                "capabilities.toml integrity mismatch, fallback all-OFF",
                path=str(path),
                stored=stored_hash,
                computed=computed,
            )
            return cls._default(path)

        caps: dict[str, CapabilityEntry] = {}
        cap_section = data.get("capability", {})
        for name in sorted(KNOWN_CAPABILITIES):
            section = cap_section.get(name, {})
            caps[name] = CapabilityEntry(
                enabled=bool(section.get("enabled", False)),
                enabled_at=str(section.get("enabled_at", "")),
                enabled_by=str(section.get("enabled_by", "")),
            )
        return cls(path, caps)

    @classmethod
    def _default(cls, path: Path) -> CapabilitiesStore:
        caps = {
            name: CapabilityEntry(enabled=False, enabled_at="", enabled_by="") for name in sorted(KNOWN_CAPABILITIES)
        }
        return cls(path, caps)

    @property
    def path(self) -> Path:
        return self._path

    def is_enabled(self, cap_name: str) -> bool:
        entry = self._capabilities.get(cap_name)
        if entry is None:
            return False
        return entry.enabled

    def enable(self, cap_name: str, actor: str) -> None:
        if cap_name not in KNOWN_CAPABILITIES:
            raise ValueError(f"unknown capability: {cap_name}")
        now = datetime.now(timezone.utc).isoformat(timespec="seconds")
        previous = self.is_enabled(cap_name)
        self._capabilities[cap_name] = CapabilityEntry(
            enabled=True,
            enabled_at=now,
            enabled_by=actor,
        )
        self._write()
        _log.info(
            "capability_changed",
            capability=cap_name,
            previous=previous,
            new=True,
            actor=actor,
        )

    def disable(self, cap_name: str) -> None:
        if cap_name not in KNOWN_CAPABILITIES:
            raise ValueError(f"unknown capability: {cap_name}")
        previous = self.is_enabled(cap_name)
        self._capabilities[cap_name] = CapabilityEntry(
            enabled=False,
            enabled_at="",
            enabled_by="",
        )
        self._write()
        _log.info(
            "capability_changed",
            capability=cap_name,
            previous=previous,
            new=False,
        )

    def get(self, cap_name: str) -> CapabilityEntry | None:
        return self._capabilities.get(cap_name)

    def audit_trail(self) -> list[dict[str, Any]]:
        return [
            {
                "capability": name,
                "enabled": entry.enabled,
                "enabled_at": entry.enabled_at,
                "enabled_by": entry.enabled_by,
            }
            for name, entry in sorted(self._capabilities.items())
        ]

    def _to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {"version": 1, "capability": {}}
        for name, entry in sorted(self._capabilities.items()):
            data["capability"][name] = {
                "enabled": entry.enabled,
                "enabled_at": entry.enabled_at,
                "enabled_by": entry.enabled_by,
            }
        return data

    def _write(self) -> None:
        data = self._to_dict()
        body = tomli_w.dumps(data)
        hash_val = "sha256-" + hashlib.sha256(body.encode()).hexdigest()
        data["integrity_hash"] = hash_val
        final = tomli_w.dumps(data)
        self._path.parent.mkdir(parents=True, exist_ok=True)
        self._path.write_text(final, encoding="utf-8")


# -- Module-level singleton for @require_capability decorator --

_store: CapabilitiesStore | None = None


def init_capabilities(path: Path | None = None) -> CapabilitiesStore:
    global _store  # noqa: PLW0603
    _store = CapabilitiesStore.load(path)
    return _store


def get_capabilities_store() -> CapabilitiesStore | None:
    return _store


def require_capability(cap_name: str):
    """FastAPI decorator — 403 when the named capability is disabled."""

    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            store = _store
            if store is None or not store.is_enabled(cap_name):
                raise HTTPException(
                    403,
                    detail=(
                        f"capability '{cap_name}' is disabled. "
                        f"Run `nexus-admin capability enable {cap_name}` "
                        "to activate."
                    ),
                )
            return await func(*args, **kwargs)

        return wrapper

    return decorator

File: crates/nexus-coordinator-rs/src/capability_store.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Capability toggles gate-off-by-default store
//! (Sprint 41 Phase B, port of capability_store.py S25).

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

pub const KNOWN_CAPABILITIES: &[&str] = &[
    "biometric_gate",
    "federation_canary",
    "mcp_server_expose",
    "rag_retrieval",
    "streaming_bridge",
    "tool_calling",
];

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CapabilityEntry {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub enabled_at: String,
    #[serde(default)]
    pub enabled_by: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CapabilityFileData {
    #[serde(default = "default_version")]
    version: u32,
    #[serde(default)]
    integrity_hash: String,
    #[serde(default)]
    capability: BTreeMap<String, CapabilityEntry>,
}

fn default_version() -> u32 {
    1
}

pub struct CapabilityStore {
    path: PathBuf,
    capabilities: BTreeMap<String, CapabilityEntry>,
}

impl CapabilityStore {
    pub fn load(path: &Path) -> Self {
        if !path.exists() {
            let store = Self::all_off(path);
            let _ = store.write();
            return store;
        }
        let text = match std::fs::read_to_string(path) {
            Ok(t) => t,
            Err(_) => return Self::all_off(path),
        };
        let data: CapabilityFileData = match toml::from_str(&text) {
            Ok(d) => d,
            Err(_) => return Self::all_off(path),
        };
        let computed = compute_integrity_hash(&text);
        if !data.integrity_hash.is_empty() && data.integrity_hash != computed {
            return Self::all_off(path);
        }
        let mut caps = BTreeMap::new();
        for &name in KNOWN_CAPABILITIES {
            let entry = data.capability.get(name).cloned().unwrap_or_default();
            caps.insert(name.to_string(), entry);
        }
        Self {
            path: path.to_path_buf(),
            capabilities: caps,
        }
    }

    pub fn all_off(path: &Path) -> Self {
        let mut caps = BTreeMap::new();
        for &name in KNOWN_CAPABILITIES {
            caps.insert(name.to_string(), CapabilityEntry::default());
        }
        Self {
            path: path.to_path_buf(),
            capabilities: caps,
        }
    }

    pub fn is_enabled(&self, cap_name: &str) -> bool {
        self.capabilities.get(cap_name).is_some_and(|e| e.enabled)
    }

    pub fn enable(&mut self, cap_name: &str, actor: &str) -> Result<(), String> {
        if !KNOWN_CAPABILITIES.contains(&cap_name) {
            return Err(format!("unknown capability: {cap_name}"));
        }
        let now = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        self.capabilities.insert(
            cap_name.to_string(),
            CapabilityEntry {
                enabled: true,
                enabled_at: now,
                enabled_by: actor.to_string(),
            },
        );
        self.write()
    }

    pub fn disable(&mut self, cap_name: &str) -> Result<(), String> {
        if !KNOWN_CAPABILITIES.contains(&cap_name) {
            return Err(format!("unknown capability: {cap_name}"));
        }
        self.capabilities
            .insert(cap_name.to_string(), CapabilityEntry::default());
        self.write()
    }

    pub fn audit_trail(&self) -> Vec<(String, bool, String, String)> {
        self.capabilities
            .iter()
            .map(|(name, entry)| {
                (
                    name.clone(),
                    entry.enabled,
                    entry.enabled_at.clone(),
                    entry.enabled_by.clone(),
                )
            })
            .collect()
    }

    fn write(&self) -> Result<(), String> {
        let data = CapabilityFileData {
            version: 1,
            integrity_hash: String::new(),
            capability: self.capabilities.clone(),
        };
        let body = toml::to_string_pretty(&data).map_err(|e| format!("TOML serialize: {e}"))?;
        let hash = compute_integrity_hash(&body);
        let mut data_with_hash = data;
        data_with_hash.integrity_hash = hash;
        let final_text =
            toml::to_string_pretty(&data_with_hash).map_err(|e| format!("TOML serialize: {e}"))?;
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent).map_err(|e| format!("create dir: {e}"))?;
        }
        std::fs::write(&self.path, final_text).map_err(|e| format!("write file: {e}"))?;
        Ok(())
    }
}

fn compute_integrity_hash(file_text: &str) -> String {
    let filtered: String = file_text
        .lines()
        .filter(|line| !line.starts_with("integrity_hash"))
        .collect::<Vec<_>>()
        .join("\n");
    format!(
        "sha256-{}",
        hex::encode(Sha256::digest(filtered.as_bytes()))
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn load_default_all_off() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("capabilities.toml");
        let store = CapabilityStore::load(&path);
        for &name in KNOWN_CAPABILITIES {
            assert!(!store.is_enabled(name));
        }
        assert!(path.exists());
    }

    #[test]
    fn enable_disable_roundtrip() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("capabilities.toml");
        let mut store = CapabilityStore::load(&path);
        assert!(!store.is_enabled("tool_calling"));
        store.enable("tool_calling", "admin").unwrap();
        assert!(store.is_enabled("tool_calling"));
        store.disable("tool_calling").unwrap();
        assert!(!store.is_enabled("tool_calling"));
    }

    #[test]
    fn integrity_hash_roundtrip() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("capabilities.toml");
        let mut store = CapabilityStore::load(&path);
        store.enable("federation_canary", "test").unwrap();
        let reloaded = CapabilityStore::load(&path);
        assert!(reloaded.is_enabled("federation_canary"));
    }

    #[test]
    fn tampered_file_falls_back_all_off() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("capabilities.toml");
        let mut store = CapabilityStore::load(&path);
        store.enable("tool_calling", "admin").unwrap();
        let mut text = std::fs::read_to_string(&path).unwrap();
        text = text.replace("enabled = true", "enabled = false");
        std::fs::write(&path, text).unwrap();
        let reloaded = CapabilityStore::load(&path);
        assert!(!reloaded.is_enabled("tool_calling"));
    }

    #[test]
    fn unknown_capability_error() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("capabilities.toml");
        let mut store = CapabilityStore::load(&path);
        assert!(store.enable("nonexistent", "admin").is_err());
    }
}
